### DISCUSSION TITLE
typedoc-plugin-markdown v4

### DISCUSSION THREAD

**Discussion by tgreyuk** [tgreyuk/typedoc-plugin-markdown#372] [Announcements]
We will be releasing an next version of typedoc-plugin-markdown v4 in the near future - a rebuild of the plugin featuring several improvements:

- Removal of handlebars 
- MDX support as standard
- Updated theming model to follow the html theme, (which means the theme can be easily overridden/adapted)
- Updated directory structure to follow the project module structure
- Ability to choose which symbols should be rendered to a single doc, with the ability to add all symbols into single doc.
- Ability to choose between list/table styles. 
- UI improvements.

Please use this area to raise feature requests and idea improvements.

**Comment by lorenzolewis** [comment]
For the list style, I have an implementation for enums specifically (you can see the render here https://tauri.app/v1/api/js/fs#basedirectory). If you aren't working on it already then I could draft up a PR in that direction.

**Reply by tgreyuk** [reply]
that looks cool - PR would be great.

**Comment by suleyman222** [comment]
Is there any timeline on when the first version of v4 will be released? The changes look great and I would love to start using it in a project.

**Reply by tgreyuk** [reply]
In a couple of weeks (hopefully). Just want to make sure all the main items in the description are covered.

**Comment by lorenzolewis** [comment]
@tgreyuk I'm starting with some low-hanging fruits. In https://github.com/tgreyuk/typedoc-plugin-markdown/pull/370 you mentioned the possibility of getting rid of dividers by default. Would you be alright with simply removing the `---` here? 

https://github.com/tgreyuk/typedoc-plugin-markdown/blob/e86bda396e03731d95956865fc4c9a83967e01ae/packages/typedoc-plugin-markdown/src/support/els.ts#L14 

Or do you have some flexibility you'd like to expose somehow to allow overriding it?

**Reply by tgreyuk** [reply]
I'd rather keep that method as a helper (as it might be useful) and just remove where its being called from the partials. If it works better without the hrs then just leave them out altogether.

**Reply by lorenzolewis** [reply]
How about just changing it from `\n\n---` to `\n\n`? Or would you like it removed altogether from the partials?

**Reply by tgreyuk** [reply]
I think just remove it from the partials. I am not even sure that we need an additional option around it. Lets just create something we think works best. If a user wants to add other stuff they can just extend the theme themselves.

**Reply by lorenzolewis** [reply]
Cool, I'll PR that

**Reply by lorenzolewis** [reply]
https://github.com/tgreyuk/typedoc-plugin-markdown/pull/382

**Comment by suleyman222** [comment]
Would be nice to see a fix for #252. I don't think there's a straightforward way to customize which characters are escaped currently. Maybe an option that looks something like ``escapedChars: ['>', '<', `{`, '_', '`', '\', ...]`` or some way to overwrite it in the context?

**Reply by tgreyuk** [reply]
Yes will look at this next.

**Reply by tgreyuk** [reply]
This should be fixed in next version typedoc-plugin-markdown@next . Please let me know .

**Reply by suleyman222** [reply]
Hey, I finally had a chance to test the `next` version and take a look at this. Thanks for the great work!

The escaping seems to be working in most cases, but I noticed one case where curly braces aren't escaped when declaring an intersection or union type with an object literal type.

For example:

```ts
export type Foo = {
    foo: string;
}

export type Bar = Foo & {
    bar: number;
}
``` 


results in:

```
### Bar

> [`Foo`](README.md#Foo) & {
    `bar`: `number`;
}
```

where the `{` isn't escaped and that leads to [this error](https://mdxjs.com/docs/troubleshooting-mdx/#could-not-parse-expression-with-acorn-unexpected-content-after-expression). This doesn't seem to be a problem for regular object literal types, since they're wrapped in code blocks.

**Reply by tgreyuk** [reply]
Thanks @suleyman222 - i have a fixed this. will let you know when its been deployed with next release.

**Comment by lorenzolewis** [comment]
Right now you're just running test suites for `Output File Strategy`:

https://github.com/tgreyuk/typedoc-plugin-markdown/blob/7a920bffb8a68c3a96fa21e6767fc4912694ff84/packages/typedoc-plugin-markdown/package.json#L15

Ideally I'd like to wait until the rest of the suites are enabled and the snapshots are where you'd like them before I start some larger PRs for list styles and the single-line return format (that would look like `Returns: Promise<string | null>`).

Is there anything I can do to help with this?

**Comment by Zamiell** [comment]
Posting here to request an update on #328, hopefully this kind of functionality can be in version 4 of the plugin.

**Reply by tgreyuk** [reply]
> Posting here to request an update on #328, hopefully this kind of functionality can be in version 4 of the plugin.

Please install @next version ( typedoc-plugin-markdown@next ).

Some docs here: https://github.com/tgreyuk/typedoc-plugin-markdown/blob/next/packages/typedoc-plugin-markdown/docs/file-output-options.md

Let me know how you get on.

**Reply by Zamiell** [reply]
Thanks Tom! I'm testing out the `next` version of the plugin now.
By default, it looks like the plugin is splitting up each function into its own individual file, like this:

![image](https://user-images.githubusercontent.com/5511220/217581671-004342d0-471d-41b4-9b48-8bf78649d666.png)

What are my options for getting around this? My ideal solution would be to specify some configuration option in the "typedoc.json" file that says "groupBySourceFile: true" or something along those lines.

(I'm really hoping I don't have to maintain a separate list of files and/or exports. It becomes very hard to maintain, because when you add a new file to the root "index.ts" in your library, you will often forget to add it to the "index.typedoc.ts" / "typedoc.json" / whatever.)

**Reply by tgreyuk** [reply]
hi @Zamiell . So currently this is the default to split everything onto own file (as per html theme). If you want to hoist all exports onto a single module file you can do `symbolsWithOwnFile: none` or configure depending on symbol type ie . `symbolsWithOwnFile: ['class', 'interface']`  Some initial docs here - https://github.com/tgreyuk/typedoc-plugin-markdown/blob/next/packages/typedoc-plugin-markdown/docs/file-output-options.md.  I am not 100% sure on the naming of this currently (and whether this is the correct default) - but this is where i am currently going. Let me know if this helps.

**Reply by Zamiell** [reply]
Thanks Tom. I tried out the `symbolsWithOwnFile: none` option, and it just packs everything together into a 33 thousand line "modules.md" file. 😆 I see that it is also possible to configure `symbolsWithOwnFile` with an array, but this doesn't help - it just controls what specific stuff gets shoved into "modules.md", and I want nothing to go there. So this is not what I need. 

Ideally, I would like each thing to be in a file corresponding to the source code, regardless of whether the "thing" is a class, interface, type, enum, etc. Would it be possible to add a new option that covers that use-case, like `groupBySourceFile: true`, or something along those lines?

In general, I think that this is an important use-case to cover, because it seems like it would be very common for people to code this way. Say that I have a library that provides 3 features: foo, bar, and baz. It just makes sense to structure the code such that all of the foo stuff is in "foo.ts", all of the bar stuff is in "bar.ts", and all of the baz stuff is in "baz.ts". That way, when someone reports a bug with foo, we don't have to go on a scavenger hunt throughout our codebase - we know that the problem must be in the file called "foo.ts". And subsequently, we want to document all of the things relating to foo in one place, on one page, so that people want to learn about foo, they don't have to wade through unrelated enums, unrelated classes, and so on.

**Reply by tgreyuk** [reply]
TypeDoc generates its module by analysing exports of entry points. I have created an example repo with 3 files foo.ts, bar.ts and baz.ts, that gets exported to 3 md files, with each file a module containing the exports of that entrypoint - essentially a one-to-one mapping of a file to a module if that is the desired outcome. You could ofcourse import and then export symbols from multiple files foo-a, foo-b if so desired.

![exports](https://user-images.githubusercontent.com/11680870/217682807-77c9569a-1fa8-4812-97ef-cd5ea1a0f96b.PNG)

Here is the repo https://github.com/tgreyuk/typedoc-plugin-markdown-module-exports-example

**Reply by Zamiell** [reply]
Thanks Tom. I took a look, and in that repo, you manually specify each file individually in the "entryPoints".

Unfortunately, this does not solve the problem, because it causes me to have to specify every single file in my entire repository inside of my "entryPoints" file, which results in hundreds of entries like this: https://github.com/IsaacScript/isaacscript/blob/main/packages/isaacscript-common/typedoc.json#L13

As I said above, this becomes very hard to maintain, because whenever we need to add a new file to the list of library exports in the "index.ts" file, we often forget to add it to the "typedoc.json" file, causing the documentation to get out of sync / become bugged.

Having to duplicate all of the information in the "index.ts" file a second time in the "typedoc.json" file violates [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), which is the root cause of these bugs. This is the motivation behind wanting an option like "groupBySourceFile".

**Reply by tgreyuk** [reply]
> Thanks Tom. I took a look, and in that repo, you manually specify each file individually in the "entryPoints".

Well, you could achieve the same thing with:

```
  "entryPoints": ["./src"],
  "entryPointStrategy": "expand",
```

**Reply by Zamiell** [reply]
Tom,

I dont think that is a solution either, because then TypeDoc would document private files that end-users should not have access to, right? I only want to document things that are specified in the "index.ts" file (the thing pointed to by the "main" field in the "package.json" file).

**Reply by tgreyuk** [reply]
right - well i am running out of ideas/suggestions im afraid. you could of course configure entrypoints by reading the file system with `path` and `fs`.

**Reply by Zamiell** [reply]
> configure entrypoints by reading the file system

That is smart; I've not thought of that before.

Looking into this more, yes, it seems like that instead of using "typedoc.json", I could instead use "typedoc.js", and then manually prepare the "entryPoints" array at the top of the file (by parsing the contents of "index.ts" file). So yes, that seems like the best option for now.

However, it would be fantastic if TypeDoc could provide an option for this out of the box (e.g. `groupBySourceFile`), as arriving at this conclusion / script setup is non-trivial.
